tsdb/db_test.go

```diff
@@ -3489,3 +3490,209 @@ func TestDBPanicOnMmappingHeadChunk(t *testing.T) {
+	app.AppendMetadata(3, labels.FromStrings("e", "f"), m1)
+	app.Commit()
+
+	// Add a replicated metadata entry to the first series,
+	// a changed metadata entry to the second series,
+	// and a completely new metadata entry for the fourth series.
+	app = db.Appender(ctx)
+	app.AppendMetadata(1, labels.FromStrings("a", "b"), m1)
+	app.AppendMetadata(2, labels.FromStrings("c", "d"), m2)
+	app.AppendMetadata(4, labels.FromStrings("g", "h"), m2)
+	app.Commit()
+
+	// Read the WAL to see if the disk storage format is correct.
+	walFiles, err := os.ReadDir(path.Join(db.Dir(), "wal"))
+	require.NoError(t, err)
+	f, err := os.OpenFile(path.Join(db.Dir(), "wal", walFiles[0].Name()), os.O_RDWR, 0o666)
+	require.NoError(t, err)
+	defer f.Close()
+
+	// The first two blocks will correspond to the encoded Series' and Samples' blocks.
+	r := wal.NewReader(bufio.NewReader(f))
+	require.True(t, r.Next(), "reading the first WAL record, it should be of type 'Series'")
+	mr := r.Record()
+	require.Equal(t, mr[0], byte(record.Series))
+
+	require.True(t, r.Next(), "reading the second WAL record, it should be of type 'Samples'")
+	mr = r.Record()
+	require.Equal(t, mr[0], byte(record.Samples))
+
+	// The next two blocks will contain the committed metadata entries.
+	// Let's decode them to see that they're correctly written.
+	var dec record.Decoder
+	var mb1 []record.RefMetadata
+	var mb2 []record.RefMetadata
+
+	require.True(t, r.Next(), "reading the third WAL record, it should be of type 'Metadata'")
+	mr = r.Record()
+	require.Equal(t, mr[0], byte(record.Metadata))
+	mb1, err = dec.Metadata(mr, mb1)
+	require.NoError(t, err)
+
+	require.True(t, r.Next(), "reading the fourth WAL record, it should be of type 'Metadata'")
+	mr = r.Record()
+	require.Equal(t, mr[0], byte(record.Metadata))
+	mb2, err = dec.Metadata(mr, mb2)
+	require.NoError(t, err)
+
+	expectedMetadata := []record.RefMetadata{
+		{Ref: 1, Type: record.GetMetricType(m1.Type), Unit: m1.Unit, Help: m1.Help},
+		{Ref: 2, Type: record.GetMetricType(m1.Type), Unit: m1.Unit, Help: m1.Help},
+		{Ref: 3, Type: record.GetMetricType(m1.Type), Unit: m1.Unit, Help: m1.Help},
+		{Ref: 2, Type: record.GetMetricType(m2.Type), Unit: m2.Unit, Help: m2.Help},
+		{Ref: 4, Type: record.GetMetricType(m2.Type), Unit: m2.Unit, Help: m2.Help},
+	}
+	require.Len(t, mb1, 3)
+	require.Len(t, mb2, 2)
+	require.Equal(t, expectedMetadata[:3], mb1)
+	require.Equal(t, expectedMetadata[3:], mb2)
+
+	require.False(t, r.Next(), "the WAL shouldn't contain any more records")
+}
+
+func TestMetadataCheckpointingOnlyKeepsLatestEntry(t *testing.T) {
+	ctx := context.Background()
+	numSamples := 10000
+	hb, w := newTestHead(t, int64(numSamples)*10, false)
+
+	// Add some series so we can append metadata to them.
+	app := hb.Appender(ctx)
+	app.Append(9001, labels.FromStrings("a", "b"), 0, 0)
+	app.Append(9002, labels.FromStrings("c", "d"), 0, 0)
+	app.Append(9002, labels.FromStrings("e", "f"), 0, 0)
```

Comment:
Same comment about using variables for the series labels.

Plus the series ref should be 0 when you don't know the ref (TSDB gives you the ref, you cannot define it). (Same for AppendMetadata). But you can assume the series ref for the expected values below and don't use variables to capture the ref here.